src/main/java/com/google/sps/servlets/utils/Statistics.java

```diff
@@ -14,29 +14,83 @@
 
 package com.google.sps.servlets.utils;
 
-import com.google.sps.servlets.Range;
+import java.math.BigDecimal;
+import java.util.Comparator;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.stream.Collectors;
 
 public class Statistics {
-  // Aggregated score for a list of comments
+  private static final double INTERVAL = 0.2;
+  private static final double UPPER_END = 1.0;
+  private static final double LOWER_END = -1.0;
+  // Contains sentiment scores in the range [-1, 1] with given intervals.
   private Map<Range, Integer> aggregateValues;
   private double averageScore;
 
+  public Map<Range, Integer> getAggregateValues() {
+    return aggregateValues;
+  }
+
+  public double getAverageScore() {
+    return averageScore;
+  }
+
   public Statistics(List<Double> sentimentScores) {
+    sentimentScores =
+        sentimentScores.stream()
+            .filter(score -> (score >= LOWER_END && score <= UPPER_END))
+            .collect(Collectors.toList());
     setAggregateScores(sentimentScores);
     setAverageScore(sentimentScores);
   }
 
+  /**
+   * Categorize all score values into different range intervals and count the frequency for each
+   * interval, and set the aggregatedValues.
+   *
+   * @param sentimentScores a list of score values from -1.0 to 1.0
+   */
   private void setAggregateScores(List<Double> sentimentScores) {
-    // TODO(Xin): Add sorting code
     aggregateValues = new HashMap<>();
-    aggregateValues.put(new Range(-1.0, 1), 1);
+    BigDecimal curPoint = BigDecimal.valueOf(LOWER_END);
+    BigDecimal interval = BigDecimal.valueOf(INTERVAL);
+
+    //     Initialize the HashMap with intervals
+    while (curPoint.doubleValue() < UPPER_END) {
+      Range currentRange =
+          new Range(
+              curPoint.doubleValue(), Math.min(curPoint.add(interval).doubleValue(), UPPER_END));
```

Comment:
Why Math.min? You could well be hitting 1.999999 and it was meant to be 2 as from http://shortn/_w5ccMdlxlu:
Note: The information about the decimal precision of the Double value of the given BigDecimal value can be lost even if the return value is finite.

Reply:
1) Cannot get access to the link http://shortn/_w5ccMdlxlu :
2) If our interval is 0.3, it reaches 0.8 -- 1.1, we need to auto change it to 0.8 -- 1.0
3) still don't know why 1.9999999999999 == 2, but it sounds like something interesting hhhh